src/rust/engine/fs/src/lib.rs

```diff
@@ -587,17 +593,27 @@ impl PosixFS {
     let dir_abs = self.root.0.join(&dir_relative_to_root.0);
     let posix_fs = self.clone();
     let ignore = self.ignore.clone();
-    tokio_fs::read_dir(dir_abs.clone())
+    self
+      .io_pool
+      .spawn_fn(move || std::fs::read_dir(&dir_abs))
       .and_then(move |readdir| {
```

Comment:
The fact that both futures and iterators implement `map` and `and_then` makes things somewhat confusing here, without an IDE. Unfortunately I don't think type annotations would help here (`.and_then(move |readdir: ReadDir|)`) is still not very iterator-y.